File: source/dotnet/source/cmf.bus/cmf.bus/support/InMemoryEnvelopeBus.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

using cmf.bus;

namespace cmf.bus.support
{
    /// <summary>
    /// An in-memory implementation of the envelope bus interface.
    /// <remarks>
    /// <para>
    /// All registrations are held in-memory and all sent envelopes are
    /// delivered only to those registrations. There is no communication 
    /// external to the process using this component, and no processing chain.
    /// </para>
    /// <para>
    /// When an envelope is 'sent', a background thread is created to dispatch
    /// the envelope to any registered consumers. That means that consumers 
    /// handle <i>an envelope</i> sequentially. It also means that consumers 
    /// could be handling multiple envelopes on multiple threads at once.
    /// </para>
    /// <para>
    /// If there are multiple registrations for an envelope and one handler 
    /// throws an exception, the exception is swallowed to ensure all handlers
    /// have an opportunity to handle the envelope.
    /// </para>
    /// <para>
    /// If you wish to change the threading or exception handling behavior, 
    /// override the Dispatch method to do whatever you want.
    /// </para>
    /// </remarks>
    /// </summary>
    public class InMemoryEnvelopeBus : IEnvelopeBus
    {
        protected List<IRegistration> _registrationList;

        private object _registrationListLock = new object();


        public InMemoryEnvelopeBus()
        {
            _registrationList = new List<IRegistration>();
        }


        public virtual void Send(Envelope env)
        {
            IEnumerable<IRegistration> handlerList = null;

            lock (_registrationListLock)
            {
                handlerList = from reg in _registrationList
                              where ((null != reg.Filter) && (reg.Filter(env)))
                              select reg;
            }

            this.Dispatch(env, handlerList);
        }

        public virtual void Register(IRegistration registration)
        {
            lock (_registrationListLock)
            {
                _registrationList.Add(registration);
            }
        }


        protected virtual void Dispatch(Envelope env, IEnumerable<IRegistration> handlerList)
        {
            Thread dispatchThread = new Thread(new ThreadStart(delegate () {
                handlerList.ToList().ForEach(handler =>
                {
                    try
                    {
                        handler.Handle(env);
                    }
                    catch (Exception ex)
                    {
                        handler.HandleFailed(env, ex);
                    }
                });
            }));

            dispatchThread.Start();
        }
    }
}

File: source/java/src/main/java/cmf/bus/support/InMemoryEnvelopeBus.java
package cmf.bus.support;

import java.util.LinkedList;
import java.util.List;

import cmf.bus.Envelope;
import cmf.bus.IEnvelopeBus;
import cmf.bus.IRegistration;

public class InMemoryEnvelopeBus implements IEnvelopeBus {

    protected List<IRegistration> registrationList;
    private Object registrationListLock = new Object();

    public InMemoryEnvelopeBus() {
        registrationList = new LinkedList<IRegistration>();
    }

    protected void dispatch(final Envelope envelope, final List<IRegistration> registrationList) {
        new Thread() {

            @Override
            public void run() {
                for (IRegistration registration : registrationList) {
                    try {
                        registration.getEnvelopeHandler().handle(envelope);
                    } catch (Exception e) {
                        registration.getEnvelopeHandler().handleFailed(envelope, e);
                    }
                }
            }
        }.run();
    }

    @Override
    public void register(IRegistration registration) {
        synchronized (registrationListLock) {
            registrationList.add(registration);
        }
    }

    @Override
    public void send(Envelope envelope) {
        List<IRegistration> registrations = new LinkedList<IRegistration>();
        synchronized (registrationListLock) {
            for (IRegistration registration : registrationList) {
                if (registration.getFilterPredicate().filter(envelope)) {
                    registrations.add(registration);
                }
            }
            dispatch(envelope, registrations);
        }
    }
}
